docs/en/03_Upgrading/02_Upgrading_module.md

```diff
@@ -0,0 +1,335 @@
+
+## Step 3 - Namespacing your module
+
+Namespacing your module is mandatory to get it working with SilverStripe 4. You can use the `add-namespace` upgrader command to achieve this.
+
+```bash
+# If you are working from a test project, you need to specify the `--root-dir` parameter
+upgrade add-namespace --root-dir vendor/example-user/silverstripe-example-module \
+  "ExampleUser\\SilverstripeExampleModule" \
+  vendor/example-user/silverstripe-example-module/code/
+  
+# If you are working directly from the module, you can ommit `--root-dir` parameter
+upgrade add-namespace "ExampleUser\\SilverstripeExampleModule" code/
+```
+
+If your module codebase is structure in folders, you can use the `--psr4` and `--recursive` flag to quickly namespace your entire module in one command. This command will recursively go trough the `code` directory and namespace all files based on their position relative to `code`.
+
+```bash
+upgrade add-namespace --recursive --psr4 "ExampleUser\\SilverstripeExampleModule" code/
+```
+
+### Configuring autoloading
+
+You need to update your `composer.json` file with an autoload entry, so composer knows what folder maps to what namespace.
+
+You can do this manually:
+```diff
+{
+    "name": "example-user/silverstripe-example-module",
+    "type": "silverstripe-vendormodule",
+    "require": {
+        "silverstripe/framework": "^4",
+        "silverstripe/vendor-plugin": "^1"
+-    }
++    },
++    "autoload": {
++        "psr4": {
++            "ExampleUser\\SilverstripeExampleModule\\": "code/"
++        }
++    },
++    "autoload-dev": {
++        "psr4": {
++            "ExampleUser\\SilverstripeExampleModule\\Tests\\": "tests/"
++        }
++    }
+}
+```
+
+Alternatively, you can use the `--autoload` or `--autoload-dev` parameter when calling `add-namespace` to do this for you.
+
+```bash
+upgrade add-namespace --recursive --psr4 --autoload "ExampleUser\\SilverstripeExampleModule" code/
+upgrade add-namespace --recursive --psr4 --autoload-dev "ExampleUser\\SilverstripeExampleModule\\Tests" tests
+```
+
+[Learn more about configuring autoloading](https://getcomposer.org/doc/04-schema.md#autoload) in your `composer.json` file.
+
+### Preparing your `.upgrade.yml` file
+
+`add-namespace` will create a `.upgrade.yml` file that maps your old class names to their new namespaced equivalent. This will be used by the `upgrade` command in the next step.
+
+Depending on the nature of your module, you may have some class names that map to other common names. When the`upgrade` command runs, it will try to substitute any occurrence of the old name with the namespaced one. This can lead to accidental substitution. For example, let's say you have a `Link` class in your module. In many project the word `Link` will be used for other purposes like a field label or property names. You can manually update your `.upgrade.yml` file to define a `renameWarnings` section. This will prompt users upgrading to confirm each substitution.
```

Comment:
```suggestion
Depending on the nature of your module, you may have some class names that map to other common names. When the `upgrade` command runs, it will try to substitute any occurrence of the old name with the namespaced one. This can lead to accidental substitution. For example, let's say you have a `Link` class in your module. In many project the word `Link` will be used for other purposes like a field label or property names. You can manually update your `.upgrade.yml` file to define a `renameWarnings` section. This will prompt users upgrading to confirm each substitution.
```